enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,193 @@
+
+#### Deleting CRs
+
+The operand cleanup process, simply put, is the removal of all CRs being reconciled by an operator so it can run finalizers for each CR and clean up any associated operator-created resources.
+
+From OLM's perspective these would be all CRs in the `OperatorGroup`'s [target namespaces][target-namespaces] that are of the same type as the [owned APIservices][owned-apiservices] listed in the CSV. This represents the set of all CRs that the operator reconciles and is responsible for cleaning up after.
+
+OLM should not delete CRs of the type [required APIService][required-apiservices] as those are likely created by the operator, and whose cleanup should be handled by the operator's finalizer.
+
+#### Finalizers on CSVs
+
+OLM can initiate the process of operand clean up when the operator is uninstalled, which is when the CSV is deleted. However, deleting a CSV causes the operator deployment (owned by the CSV) to be garbage collected.
+
+So in order to keep the operator around to process CR deletion events, OLM can add a finalizer to the CSV to prevent the CSV and the operator from being removed. Once OLM has completed operand cleanup and can verify the removal of all CRs, the CSV finalizer can be cleared so that the operator can be safely removed.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+metadata:
+  ...
+  finalizers:
+  - olm/operand-cleanup
+```
+
+In the event of operand cleanup being stalled due to some operator error, OLM will not clear the CSV finalizer so that the operator stays alive indefinitely. This way the operator can be observed to figure out the required action to unblock cleanup.
+
+#### Operand cleanup API
+
+The CSV spec can have an optional field `spec.cleanup` which lets a user configure the cleanup behaviour before the CSV is deleted.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    enabled: false
+```
+
+With `spec.cleanup.enabled: false` OLM will not perform cleanup or delete any CRs. The finalizer on the CSV will be cleared so the CSV and operator can be removed without any cleanup. This is also the default behavior if `spec.cleanup` is unspecified to make cleanup opt-in.
+
+With `cleanup.enabled: true` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces of that operator. Once the CRs are successfully removed (after the operator clears its finalizers), OLM will clear the finalizer from the CSV to let the operator be uninstalled.
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
```

Comment:
I would be concerned about scalability here. Since the number of CRs is not bounded, committing to listing _all_ of them in the CSV status could theoretically make the CSV object large enough to break things (like etcd). Granted this is a pretty unlikely edge case -- is it an acceptable risk?

Reply:
While I haven't done the math to figure out the exact CSV size that would break the etcd storage limit (of 3MB I think), this could be a risk.

Since the intent of listing the CRs on the status is to indicate that something is still pending, I think ordering the list to just the first N CRs could be a reasonable way to avoid blowing up the CSV size. 
Where N is something that's useful at a quick glance e.g 20. 
For the complete list you could always resort to `kubectl get`.

Reply:
Sounds reasonable -- as long as it can be made reasonably obvious that it's a partial list. Not sure how you would do that.

Reply:
> Since the intent of listing the CRs on the status is to indicate that something is still pending, I think ordering the list to just the first N CRs could be a reasonable way to avoid blowing up the CSV size.

This sounds reasonable, but I also think we can get away with something more coarse, like a counter.

I think we also need status condition saying "cleanup in progress".